client/windows/CMessage.cpp

```diff
@@ -60,125 +60,227 @@ void CMessage::dispose()
-			symbolSize = TextOperations::getUnicodeCharacterSize(text[currPos]);
-
-			// candidate for line break
-			if(ui8(text[currPos]) <= ui8(' '))
-				wordBreak = currPos;
+			logGlobal->error("CMessage: Line validation Invalid: Closing tag before opening tag %s", text);
+			return false;
+		}
+		if(openingPos < closingPrevPos)
+		{
+			logGlobal->error("CMessage: Line validation Invalid: Tags inside another tags %s", text);
+			return false;
+		}
+		closingPrevPos = *closingPos++;
+	}
+	return true;
+}
 
-			/* We don't count braces in string length. */
-			if(text[currPos] == '{')
+std::vector<CMessage::coloredline> CMessage::getPossibleLines(
+	const std::string & line, size_t maxLineWidth, const EFonts font, const char & splitSymbol)
+{
+	const auto & fontPtr = GH.renderHandler().loadFont(font);
+	std::vector<coloredline> result;
+	std::string_view lastFoundColor;
+	const std::function<size_t(const std::string_view&)> findColorTags = [&](const std::string_view & subLine) -> size_t
+		{
+			size_t foundWidth = 0;
+			if(lastFoundColor.empty())
 			{
-				opened = true;
-
-				std::smatch match;
-				std::regex expr("^\\{(.*?)\\|");
-				std::string tmp = text.substr(currPos);
-				if(std::regex_search(tmp, match, expr))
+				if(const auto colorTextBegin = subLine.find('{'); colorTextBegin != std::string::npos)
 				{
-					std::string colorText = match[1].str();
-					if(auto c = Colors::parseColor(colorText))
+					foundWidth += fontPtr->getStringWidth("{");
+					// Opened color tag brace found. Try to find color tag.
+					if(const auto colorTagEnd = subLine.find('|'); colorTagEnd != std::string::npos)
+					{
+						assert(subLine.find('}') > colorTagEnd);
+						lastFoundColor = subLine.substr(colorTextBegin, colorTagEnd - colorTextBegin + 1);
+						foundWidth += fontPtr->getStringWidth(std::string(lastFoundColor));
+					}
+					else
 					{
-						color = colorText + "|";
-						currPos += colorText.length() + 1;
+						lastFoundColor = subLine.substr(colorTextBegin, 1);
 					}
 				}
+				else
+				{
+					return foundWidth;
+				}
 			}
-			else if(text[currPos] == '}')
+			if(const auto colorTextEnd = subLine.find('}'); colorTextEnd != std::string::npos)
 			{
-				opened = false;
-				color = "";
+				foundWidth += fontPtr->getStringWidth("}");
+				lastFoundColor = std::string_view();
+				foundWidth += findColorTags(subLine.substr(colorTextEnd + 1, subLine.length() - colorTextEnd - 1));
 			}
-			else
-				printableString.append(text.data() + currPos, symbolSize);
-			currPos += symbolSize;
-		}
+			return foundWidth;
+		};
 
-		// not all line has been processed - it turned out to be too long, so erase everything after last word break
-		// if string consists from a single word (or this is Chinese/Korean) - erase only last symbol to bring line back to allowed length
-		if(currPos < text.length() && (text[currPos] != 0x0a))
+	const std::function<void(const std::string_view&)> splitBySymbol = [&](const std::string_view & subLine)
```

Comment:
Same here

Reply:
This lambda doesn't make sense outside of getPossibleSublines. I think adding an extra entity will make it harder to understand the code logic.